### PR TITLE
RFC: Remove support for MSVC

### PR DESCRIPTION
**Pull request by michael-schwarz**
We have not tested any of our modifications to CIL with MSVC, and afaik have no intention of adding support for MSVC to Goblint.

This PR removes all support for MSVC from CIL, allowing us to make some things less complex inside CIL.
It also allows us to remove the `skind`s associated with the proprietary Microsoft try/catch extension, which is not supported by Goblint. This will allow some simplifications in Goblint as well. 

I feel it is better to remove all MSVC things than having them around without ever testing if they still work and having no intention of doing so in the future. WDYT?

### CODE REVIEW

File: lib/perl5/patcher [hunk 1]
```diff
@@ -66,7 +66,6 @@ options:
   --verbose    Prints a lot of information about what is being done
   --mode=xxx   What tool to emulate: 
                 GNUCC   - GNU CC
-                MSVC    - MS VC cl compiler
                 EDG     - EDG front end
```
Comment (sim642) [inline lib/perl5/patcher]: What's this EDG that's still supported here?

File: lib/perl5/patcher [hunk 2]
```diff
@@ -204,44 +200,28 @@ exit $hadError;
...
-                return;
-            }
+    $cname = "GNU CC";
+    open(VER, "$::cc -dumpversion $ppargs|") 
+        || die "Cannot start $cname";
+    while(<VER>) {
+        # sm: had to modify this to match "egcs-2.91.66", which is
+        # how egcs responds to the -dumpversion request
+        if($_ =~ m|^(\d+\S+)|  ||
+            $_ =~ m|^(egcs-\d+\S+)|) {
```
Comment (sim642) [inline lib/perl5/patcher]: Also some EDG stuff here?

File: src/cil.mli [hunk 3]
```diff
@@ -1442,9 +1416,6 @@ val unrollType: typ -> typ
  * types. Will collect all attributes *)
 val unrollTypeDeep: typ -> typ
 
-(** Separate out the storage-modifier name attributes *)
-val separateStorageModifiers: attribute list -> attribute list * attribute list
```
Comment (sim642) [inline src/cil.mli:1446]: Why is this removed?

File: src/frontc/clexer.mll [hunk 4]
```diff
@@ -562,9 +548,7 @@ rule initial =
 |		','				{COMMA}
 |		'.'				{DOT}
 |		"sizeof"		{SIZEOF (currentLoc ())}
-|               "__asm"                 { if !Cprint.msvcMode then
-                                             MSASM (msasm lexbuf, currentLoc ())
```
Comment (sim642) [inline src/frontc/clexer.mll:566]: The `MSASM` (and `MSATTR`) constructors in `Cparser` could now be removed, no?

File: src/cil.ml [hunk 5]
```diff
@@ -1444,26 +1414,26 @@ type attributeClass =
...
       "exception"; "model"; (* "restrict"; *)
       "aconst"; "__asm__" (* Gcc uses this to specify the name to be used in
                            * assembly for a global  *)];
 
   (* Now come the MSVC declspec attributes *)
-  List.iter (fun a -> H.add table a (AttrName true))
+  List.iter (fun a -> H.add table a (AttrName))
     [ "thread"; "naked"; "dllimport"; "dllexport";
       "selectany"; "allocate"; "nothrow"; "novtable"; "property";  "noreturn";
       "uuid"; "align" ];
```
Comment (sim642) [inline src/cil.ml]: The comment says "MSVC declspec". Are these still relevant for GCC or what?

**Review by sim642 — COMMENTED** [review]
This probably doesn't make too much of a difference for Goblint, since we already error on `TryExcept`/`TryFinally` during CFG construction.

But it's also true that we haven't ever tested or maintained this, so it's better to not pretend either.

File: src/cil.ml (hunk 5 above)
Comment (michael-schwarz) [reply src/cil.ml]: Yes, GCC supports (some of) these as well, but does not require them to be printed in a special way like MSVC does.

File: src/cil.mli (hunk 3 above)
Comment (michael-schwarz) [reply src/cil.mli:1446]: Without support for MSVC there are no special attributes anymore that must be printed different from others. Therefore, the first list is always empty, making the function unnecessary.

File: src/frontc/clexer.mll (hunk 4 above)
Comment (michael-schwarz) [reply src/frontc/clexer.mll:566]: Yup, I removed them. This means that `tests/small1/attr.c` no longer can be parsed, but GCC rejects the program as well.

File: lib/perl5/patcher (hunk 1 above)
Comment (michael-schwarz) [reply lib/perl5/patcher]: It is some strange frontend from the Edison Design Group that is apparently used by ICC and some other compilers. I think I'll just remove it as well.

https://www.edg.com/c

File: lib/perl5/patcher (hunk 2 above)
Comment (michael-schwarz) [reply lib/perl5/patcher]: No, it is EGCS, which apparently was some fork of GCC that was reunified with GCC in April 1999. https://gcc.gnu.org/wiki/History#EGCS

But I guess it is safe to remove that too.

Comment (michael-schwarz) [reply lib/perl5/patcher]: I dropped support for EGCS as well. No one will use CIL with a compiler that became obsolete in 1999.

**Review by sim642 — APPROVED** [review]

File: src/cil.ml [hunk 6]
```diff
@@ -1428,14 +1404,8 @@ let hasAttribute s al =
...
-    AttrName of bool
-        (* Attribute of a name. If argument is true and we are on MSVC then
-         * the attribute is printed using __declspec as part of the storage
-         * specifier  *)
-  | AttrFunType of bool
-        (* Attribute of a function type. If argument is true and we are on
-         * MSVC then the attribute is printed just before the function name *)
-
+    AttrName  (* Attribute of a name. *)
+  | AttrFunType  (* Attribute of a function type. *)
```
Comment (sim642) [inline src/cil.ml:1408]: Also here you could remove all the now unnecessary parenthesis around whereever these constructors are used.

File: src/cil.ml [hunk 7]
```diff
@@ -1909,27 +1876,6 @@ let getParenthLevelAttrParam (a: attrparam) =
...
 
-(* Separate out the storage-modifier name attributes *)
-let separateStorageModifiers (al: attribute list) =
-  let isstoragemod (Attr(an, _): attribute) : bool =
-    try
-      match H.find attributeHash an with
-        AttrName issm -> issm
-      | _ -> false
-    with Not_found -> false
-  in
```
Comment (jerhard) [inline src/cil.ml:1920]: ~~As `attributeHash` is only filled with entry values `AttrName false`, this only returns the second list,~~ (see below) which was the reason why you removed it when there's only GCC.
It seems though that previously there was the attempt to handle the storage modifiers separately, but somehow it was forgotten to handle this properly?

Comment (michael-schwarz) [reply src/cil.ml:1920]: Are there any additional storage modifiers other than the storage class and whether something is inline? These are not part of the attributes here.

Any attributes that are associated with a type are also printed before the name of the variable as part of `self#pType`.

```
and storage =
    NoStorage                         (** The default storage. Nothing is  printed  *)
    | Static
    | Register
    | Extern
```

https://github.com/goblint/cil/blob/b6d354f69f4d1c85c78734fff32c3ed9ce917a78/src/cil.ml#L3288-L3296

So the question here is: Do we have any attributes associated with a variable that need to be printed before its name? I could not find any such case when going over our attributes, but I may very well have missed something,

File: src/cil.ml [hunk 8]
```diff
@@ -1444,26 +1414,26 @@ type attributeClass =
...
       "exception"; "model"; (* "restrict"; *)
       "aconst"; "__asm__" (* Gcc uses this to specify the name to be used in
                            * assembly for a global  *)];
 
   (* Now come the MSVC declspec attributes *)
-  List.iter (fun a -> H.add table a (AttrName true))
+  List.iter (fun a -> H.add table a AttrName)
     [ "thread"; "naked"; "dllimport"; "dllexport";
       "selectany"; "allocate"; "nothrow"; "novtable"; "property";  "noreturn";
       "uuid"; "align" ];
```
Comment (jerhard) [inline src/cil.ml]: These were previously separated by `separateStorageModifiers` in front, as they were mapped to `AttrName true`, right?

Comment (michael-schwarz) [reply src/cil.ml]: Yes, but it seems they need not be separated on GCC.

File: src/cil.ml (hunk 7 above)
Comment (jerhard) [reply src/cil.ml:1920]: So these storage class attributes are a [MSVC specific thing](https://docs.microsoft.com/en-us/cpp/c-language/c-extended-storage-class-attributes?view=msvc-170). 

But I first misread the code that initalized the `attributeHash` hashtable, there actually were some keys that were associated with `AttrName true`. (See my other comment.) So these were previously printed before the name.

File: src/cil.ml (hunk 8 above)
Comment (jerhard) [reply src/cil.ml]: Are all of these MSVC-specific? C11 introduced also a `noreturn` / `_Noreturn`, but this is probably not on the base branch?

Comment (jerhard) [reply src/cil.ml]: If they are MSVC specific, one should remove them probably.

Comment (michael-schwarz) [reply src/cil.ml]: I removed the ones that are MSVC specific now. GCC picked up support for a lot of them over the years for compatibility purposes. 

`noreturn` is in fact supported via this, it has also been a GCC extension for a long time. In the C11 branch, we reuse this pre-C11 support to handle it.

Comment (jerhard) [reply src/cil.ml]: Good! I think then the comment above: `   (* Now come the MSVC declspec attributes *)` is obsolete or should be changed to indicate that this is still needed for GCC?

Comment (jerhard) [reply src/cil.ml]: Good! I think then the comment above: ` (* Now come the MSVC declspec attributes *)` is obsolete or should be changed to indicate that this is still needed for GCC?

**Review by jerhard — APPROVED** [review]
